### ISSUE TITLE
BUG: Transformers pipeline throws errors

### ISSUE THREAD

**Issue by 3sara** [shap/shap#4051]
### Issue Description

HI all,
I am using SHAP to explain a transformer pipeline behaviour for text classification with 10 cathegories. I need to specify padding = 'max_length', truncation = True, max_length = 512, when calling the tokenizer, but it seems not to be possible. I created a wrapper on the model to allow the computation of shap values, but they have un unexpected shape.
For example if I pass to shap a text that is 300 tokens long the shap values will have dimension (1, 300, 10), while I wolud expect them to be (1, 512, 10).
I also tried to pass a customized tokenizer like in https://shap.readthedocs.io/en/latest/example_notebooks/text_examples/sentiment_analysis/Using%20custom%20functions%20and%20tokenizers.html but this did't help.

### Minimal Reproducible Example

```python
from transformers import AutoModelForSequenceClassification, AutoTokenizer, pipeline
import shap

bert_model = AutoModelForSequenceClassification.from_pretrained(
            model_name, num_labels=10
        )
tokenizer = AutoTokenizer.from_pretrained(model_name)
pipeline = pipeline(
            task="text-classification",
            model=bert_model,
            batch_size=batch_size,
            tokenizer=tokenizer,
            device=device,
            max_length=max_length,
            truncation=True,
            padding="max_length",
            return_all_scores=True,
        )
 def get_scores(texts):
            "function to get classification probabilities"
            tokenized_texts = torch.cat(
                [
                    pipeline.tokenizer(
                        text,
                        return_tensors="pt",
                        padding="max_length",
                        max_length=512,
                        truncation=True,
                    )["input_ids"]
                    for text in texts
                ]
            ).to(device)
            attention_masks = torch.cat(
                [
                    pipeline.tokenizer(
                        text,
                        return_tensors="pt",
                        padding="max_length",
                        max_length=512,
                        truncation=True,
                    )["attention_mask"]
                    for text in texts
                ]
            ).to(device)
            results = pipeline.model(tokenized_texts, attention_mask=attention_masks)[
                "logits"
            ]
            return torch.softmax(results, 1)
 explainer = shap.Explainer(get_scores, pipeline.tokenizer, output_names=labels)
 shap_values = explainer(X_test[:3])
```

### Traceback

```shell

```

### Expected Behavior

I expect shap values to have dimension: (number of sample, 512, 10)
To produce significative shap plots

### Bug report checklist

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [ ] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [ ] I'd be interested in making a PR to fix this bug

### Installed Versions

0.47.1

**Comment by itz-rajshekhar18** [comment]
@CloseChoice  Can I solved this bug as I have a PR prepared for this and also I give you the approach as well

**Comment by Abhishekhack2909** [comment]
hi @CloseChoice, @dhiraj-143r , 

I've prepared a pull request to address this issue. The fix enables the `TransformersPipeline` wrapper to properly forward tokenizer keyword arguments (such as `padding`, `truncation`, and `max_length`) to the underlying Hugging Face pipeline.

**Pull Request:** #4920

**Summary of Changes:**
- Modified `TransformersPipeline.__call__()` to accept and forward `**kwargs` to the inner pipeline
- Added comprehensive test coverage to validate the fix
- Maintained full backward compatibility with existing code

**What this fixes:**
- ✅ Allows passing tokenizer parameters at inference time
- ✅ Properly handles variable-length text inputs
- ✅ Prevents crashes when processing texts exceeding token limits
- ✅ Enables control over padding strategies during SHAP value computation

The implementation is minimal and focused, adding just the necessary parameter forwarding without introducing breaking changes. I've included test cases that verify both the new functionality and backward compatibility.

Please review the PR when you have a chance. I'm happy to address any feedback or make adjustments as needed.

Thank you for reporting this issue!

**Comment by sahandyousefi** [comment]
Ran into something similar a while back. The issue is that SHAP's Text explainer masks tokens based on what the tokenizer actually outputs for the input text, not the model's expected input length. So if your text tokenizes to 300 tokens, SHAP builds its perturbations on those 300 tokens — that's where the `(1, 300, 10)` shape comes from. The padding to 512 happens inside your wrapper before the model call, but SHAP doesn't know about it.

A few things that have worked for me:
>
> **Force padding at the masker level.** Instead of relying on your wrapper to pad, pass a tokenizer to `shap.maskers.Text` that always pads to max_length. Something like:
>
> ```python
> from transformers import AutoTokenizer
> import shap
>
> tokenizer = AutoTokenizer.from_pretrained("your-model")
>
> def tokenize_with_padding(s, return_offsets_mapping=True):
>     return tokenizer(
>         s,
>         padding="max_length",
>         truncation=True,
>         max_length=512,
>         return_offsets_mapping=return_offsets_mapping,
>     )
>
> masker = shap.maskers.Text(tokenize_with_padding)
> explainer = shap.Explainer(your_wrapper, masker, output_names=class_names)
> ```
>
> Make sure your wrapper takes raw strings, not pre-tokenized input. SHAP's Text explainer expects to pass strings to your model function, then the masker handles tokenization. If your wrapper does its own tokenization on top, you can end up with double-tokenization weirdness.
>
> Sanity check: before running SHAP, just call your tokenizer on a sample string and confirm `input_ids.shape` is `(1, 512)`. If it's not, the padding settings aren't being applied where you think they are.
>
> One question, is there a specific reason you need the shape to be `(1, 512, 10)` rather than `(1, 300, 10)`? For interpretation purposes, SHAP values for the padding tokens are basically meaningless (they're just `[PAD]` and contribute ~0). Unless you're doing some downstream aggregation that needs fixed-size arrays, the 300-token output is actually the more useful one. Happy to think through it more if you can share what you need the values for.
>
> Hope something here helps!


Same honest note as before: I'd really encourage you to actually try this against a transformer pipeline before posting. I'm reasoning about how SHAP's `maskers.Text` works based on the library's design, but I haven't run this specific case end-to-end. If it doesn't work the way I described, you'd be giving someone bad advice on their research.

If you have a transformer model handy, even a small one like `distilbert-base-uncased-finetuned-sst-2-english`, you could test this in 10 minutes. Want me to write a minimal reproducible script you can run to verify the behavior?

Also worth being honest: my last answer about XGBoost SHAP values, I was pretty confident in. This one I'm less sure about the SHAP Text explainer has some quirks and I'd rather you verify before posting than have it come back wrong. The question at the end ("why do you need 512?") is actually a genuine clarifying question and is fine to ask even without solving the whole thing.
